testutil/validatormock/windowman.go

```diff
@@ -0,0 +1,287 @@
+
+	// Manage epoch state on startup or in the first slot of an epoch.
+	if isStartup || slot.FirstInEpoch() {
+		m.manageEpochState(slot.Epoch())
+	}
+
+	// Get duties to perform this slot
+	duties := m.dutiesForSlotAndTypes(slot, core.AllDutyTypes()...)
+
+	// If startup, add startup duty types for each of their lookback slots.
+	if isStartup {
+		for _, lookbackSlot := range slot.Epoch().SlotsForLookBack(epochWindow) {
+			for duty := range m.dutiesForSlotAndTypes(lookbackSlot, startupLookbackDuties...) {
+				duties[duty] = struct{}{}
+			}
+		}
+	}
+
+	for _, duty := range orderByTime(duties) {
+		m.scheduled <- duty
+	}
+}
+
+func (m *dutyWindowManager) manageEpochState(epoch metaEpoch) {
+	// Delete attesters for the previous epoch.
+	m.deleteAttesters(epoch.Prev())
+
+	// TODO: Also delete syncCommMembers
+
+	// Start attesters for this up to lookAhead epoch if not present (idempotent).
+	for i := 0; i < epochWindow; i++ {
+		m.startAttesters(epoch)
+
+		// // TODO: Also start syncCommMembers
+
+		epoch = epoch.Next()
+	}
+}
+
+// runDuty is called to execute the duty at the appropriate time.
+func (m *dutyWindowManager) runDuty(ctx context.Context, duty core.Duty) error {
+	attester := m.slotAttester(duty.Slot)
+	syncComm := m.syncCommMember(duty.Slot)
+
+	eth2Slot := eth2p0.Slot(duty.Slot)
+
+	switch duty.Type {
+	case core.DutyPrepareAggregator:
+		attester.Prepare(ctx)
+	case core.DutyAttester:
+		attester.Attest(ctx)
+	case core.DutyAggregator:
+		attester.Aggregate(ctx)
+	case dutySubscribeSyncContribution:
+		syncComm.PrepareEpoch(ctx) // Rename to sync.Comm.SubscribeSyncContribution
+	case core.DutyPrepareSyncContribution:
+		syncComm.PrepareSlot(ctx, eth2Slot) // Rename to sync.Comm.SelectSyncContribution
+	case core.DutySyncMessage:
+		syncComm.Message(ctx, eth2Slot) // Rename to sync.Comm.SyncMessage
+	case core.DutySyncContribution:
+		syncComm.Aggregate(ctx, eth2Slot) // Rename to sync.Comm.AggregateSyncContribution
+	}
+
+	return nil
+}
+
+func (m *dutyWindowManager) startAttesters(epoch metaEpoch) {
+	for _, slot := range epoch.Slots() {
+		attester := NewSlotAttester(m.eth2Cl, eth2p0.Slot(slot.Slot), m.signFunc, m.pubkeys)
+		m.setAttester(attester)
+	}
+}
+
+func (m *dutyWindowManager) deleteAttesters(epoch metaEpoch) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, slot := range epoch.Slots() {
+		delete(m.attestersBySlot, slot.Slot)
+	}
+}
+
+func (m *dutyWindowManager) slotAttester(slot int64) *SlotAttester {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.attestersBySlot[slot] // Make nil values valid noops
+}
+
+func (m *dutyWindowManager) syncCommMember(slot int64) *SyncCommMember {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.syncCommsBySlot[slot]
+}
+
+func (m *dutyWindowManager) isStartup() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	isStartup := !m.started
+	return isStartup
+}
+
+func (m *dutyWindowManager) setAttester(attester *SlotAttester) {
+	m.mu.Lock()
+	m.mu.Unlock()
```

Comment:
It seems there's a missing `defer` statement after acquiring the mutex lock. 
```
m.mu.Lock()
m.mu.Unlock()
```
immediately locks and then unlocks the mutex, which essentially means the mutex does nothing useful in this context